### PR TITLE
Add backward compatibility for elasticsearch<8

### PR DESCRIPTION
**Pull request by sunank200**
For elasticsearch>8, arguments like `retry_timeout` has changed for elasticsearch to `retry_on_timeout` in `Elasticsearch()` compared to previous versions. Read more at: [documentation](https://elasticsearch-py.readthedocs.io/en/v8.8.2/api.html#module-elasticsearch) This change was done as part of the following [PR](https://github.com/apache/airflow/pull/33135).

This needs backward compatibility support for elasticsearch<8. It fails with following error otherwise:
```
wait-for-airflow-migrations Unable to load the config, contains a configuration error.
wait-for-airflow-migrations Traceback (most recent call last):
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/logging/config.py", line 565, in configure
wait-for-airflow-migrations     handler = self.configure_handler(handlers[name])
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/logging/config.py", line 746, in configure_handler
wait-for-airflow-migrations     result = factory(**kwargs)
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/site-packages/airflow/providers/elasticsearch/log/es_task_handler.py", line 104, in __init__
wait-for-airflow-migrations     self.client = elasticsearch.Elasticsearch(host, **es_kwargs)  # type: ignore[attr-defined]
wait-for-airflow-migrations TypeError: Elasticsearch.__init__() got an unexpected keyword argument 'retry_timeout'
wait-for-airflow-migrations 
wait-for-airflow-migrations The above exception was the direct cause of the following exception:
wait-for-airflow-migrations 
wait-for-airflow-migrations Traceback (most recent call last):
wait-for-airflow-migrations   File "/usr/local/bin/airflow", line 5, in <module>
wait-for-airflow-migrations     from airflow.__main__ import main
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/site-packages/airflow/__init__.py", line 68, in <module>
wait-for-airflow-migrations     settings.initialize()
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/site-packages/airflow/settings.py", line 524, in initialize
wait-for-airflow-migrations     LOGGING_CLASS_PATH = configure_logging()
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/site-packages/airflow/logging_config.py", line 74, in configure_logging
wait-for-airflow-migrations     raise e
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/site-packages/airflow/logging_config.py", line 69, in configure_logging
wait-for-airflow-migrations     dictConfig(logging_config)
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/logging/config.py", line 811, in dictConfig
wait-for-airflow-migrations     dictConfigClass(config).configure()
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/logging/config.py", line 572, in configure
wait-for-airflow-migrations     raise ValueError('Unable to configure handler '
wait-for-airflow-migrations ValueError: Unable to configure handler 'task'
wait-for-airflow-migrations Exception ignored in atexit callback: <function shutdown at 0x7fd39b485870>
wait-for-airflow-migrations Traceback (most recent call last):
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/logging/__init__.py", line 2183, in shutdown
wait-for-airflow-migrations     h.close()
wait-for-airflow-migrations   File "/usr/local/lib/python3.10/site-packages/airflow/providers/elasticsearch/log/es_task_handler.py", line 366, in close
wait-for-airflow-migrations     if not self.mark_end_on_close or getattr(self, "ctx_task_deferred", None):
wait-for-airflow-migrations AttributeError: 'ElasticsearchTaskHandler' object has no attribute 'mark_end_on_close'
```

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/elasticsearch/log/es_task_handler.py [hunk 1]
```diff
@@ -98,6 +98,15 @@ def __init__(
...
     ):
         es_kwargs = es_kwargs or {}
+        # For elasticsearch>8,arguments like retry_timeout have changed for elasticsearch to retry_on_timeout
+        # in Elasticsearch() compared to previous versions.
+        # Read more at: https://elasticsearch-py.readthedocs.io/en/v8.8.2/api.html#module-elasticsearch
+        if es_kwargs:
+            retry_timeout = es_kwargs.get("retry_timeout")
+            if retry_timeout:
+                es_kwargs["retry_on_timeout"] = retry_timeout
+                del es_kwargs["retry_timeout"]
```
Comment (utkarsharma2) [inline airflow/providers/elasticsearch/log/es_task_handler.py:108]: @sunank200 I think we need to be backward compatible with elasticsearch=7 as well right?

Comment (sunank200) [reply airflow/providers/elasticsearch/log/es_task_handler.py:108]: yes @utkarsharma2 I am pushing the change now

File: airflow/providers/elasticsearch/log/es_task_handler.py [hunk 2]
```diff
@@ -98,6 +98,15 @@ def __init__(
         log_id_template: str | None = None,
     ):
         es_kwargs = es_kwargs or {}
+        # For elasticsearch>8,arguments like retry_timeout have changed for elasticsearch to retry_on_timeout
+        # in Elasticsearch() compared to previous versions.
+        # Read more at: https://elasticsearch-py.readthedocs.io/en/v8.8.2/api.html#module-elasticsearch
+        if es_kwargs:
+            retry_timeout = es_kwargs.get("retry_timeout")
```
Comment (pankajkoti) [inline airflow/providers/elasticsearch/log/es_task_handler.py]: IMO, https://github.com/apache/airflow/pull/33135 should be marked as a breaking change and released with a major bump. This is one argument that we have found, but there could be more such arguments and adding back compat for them will be challenging and will be discovered only when users face issues based on the params they are using.

Comment (pankajkoti) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: cc: @Owen-CH-Leung @potiuk @eladkal WDYT?

File: chart/values.yaml [hunk 3]
```diff
@@ -2146,7 +2146,7 @@ config:
   elasticsearch_configs:
     max_retries: 3
     timeout: 30
-    retry_timeout: 'True'
+    retry_on_timeout: 'True'
```
Comment (utkarsharma2) [inline chart/values.yaml]: Should we change this? as it will fail for elasticsearch<=7 right? I think we can avoid this change now that we have a fix.

Comment (sunank200) [reply chart/values.yaml]: Yes, we should because in providers.yaml it points to >8 https://github.com/apache/airflow/blob/main/airflow/providers/elasticsearch/provider.yaml#L56

File: airflow/providers/elasticsearch/log/es_task_handler.py (hunk 2 above)
Comment (potiuk) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: I think it indeed HAS potential of breaking things. If we can make it "reasonably compatible" - i.e. fix back-compatibilities that we know and smooth the migration, that would be great, but I would be for marking the next ES release as MAJOR regardless cc: @eladkal 

Also there is a little twist to it. Unless I am mistaken, I think elasticsearch handler integration was anyhow broken and not working before :) . So well. it could also be seen as bugfix :P

File: chart/values.yaml (hunk 3 above)
Comment (utkarsharma2) [reply chart/values.yaml]: Ya, but while using someone can install the old provider version

Comment (sunank200) [reply chart/values.yaml]: reverted it

File: airflow/providers/elasticsearch/log/es_task_handler.py (hunk 2 above)
Comment (utkarsharma2) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: I too think we do a major release since the underline dependency had also changed from 7.* to 8.*

Comment (pankajkoti) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: Okay. yes +1 to make backward-compatible as much as possible.

I am not aware of how far it was broken. But, we have few tests and users for whom the elasticsearch handler integration was known to be working well. However, only yesterday our tests caught that the PR broke existing working setup :) 

Perhaps our QA expert @vatsrahul1001 can provide more confirmation here.

Comment (eladkal) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: https://github.com/apache/airflow/pull/33135 did not break any airflow code.

We agreed that bumping dependency is not a breaking change and as expected every time you bump x.y.z to x+1.y.z something can be broken for users.

I am not convencied next release should be a major one.

Comment (ephraimbuddy) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: @eladkal , I think if our helm chart can't work with the new Es provider then it should be considered as breaking change. 
Also, maybe we should consider bumping to major versions whenever we bump dependencies to major versions?

Comment (utkarsharma2) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: @eladkal I'm not entirely sure, but the below code was part of airflow's codebase which led to the issue because of upgrade - https://github.com/apache/airflow/blob/01a6c1e7f8b0ed736d7c6f661e0c29ec023016b2/chart/values.yaml#L2149

Comment (pankajkoti) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: yes, I agree bumping some dependency from x.y.z to x+1.y.z may not need to be a breaking change in general. 👍🏽 

However, would like to bring up and discuss the following point. 
For this provider, elasticsearch is the main underlying dependency, and we're updating it to the next major version from 7.x to 8.x. It does not affect core Airflow code/functionality, but since providers are versioned and released independently, upgrading to this provider release might affect existing deployments for users.

Comment (eladkal) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: @ephraimbuddy I didn't see helm tests were broken?
and to my understanding this PR brings back support for elastic search 7 so what is the motivation for the major release? Once this PR is merged the code is backward compatible.

> Also, maybe we should consider bumping to major versions whenever we bump dependencies to major versions?

This was discussed and had lazy consequence if I remember correctly. Need to lookup the thread.

File: airflow/providers/elasticsearch/log/es_task_handler.py [hunk 4]
```diff
@@ -98,6 +98,15 @@ def __init__(
         log_id_template: str | None = None,
     ):
         es_kwargs = es_kwargs or {}
+        # For elasticsearch>8,arguments like retry_timeout have changed for elasticsearch to retry_on_timeout
+        # in Elasticsearch() compared to previous versions.
+        # Read more at: https://elasticsearch-py.readthedocs.io/en/v8.8.2/api.html#module-elasticsearch
```
Comment (ephraimbuddy) [inline airflow/providers/elasticsearch/log/es_task_handler.py:103]: if es_kwargs.get("retry_timeout"):
    es_kwargs["retry_on_timeout"] = es_kwargs.pop("retry_timeout")

I think the logic should be like this

File: airflow/providers/elasticsearch/log/es_task_handler.py [hunk 5]
```diff
@@ -126,6 +135,27 @@ def __init__(
         self._doc_type_map: dict[Any, Any] = {}
         self._doc_type: list[Any] = []
 
+    @staticmethod
+    def format_url(host: str) -> str:
+        """
+        Formats the given host string to ensure it starts with 'http' or 'https'.
```
Comment (pankajkoti) [inline airflow/providers/elasticsearch/log/es_task_handler.py]: I think we need to rephrase this as we only try prefixing `http` and not `https` if the URL does not have a scheme.

Comment (sunank200) [reply airflow/providers/elasticsearch/log/es_task_handler.py]: changed it: https://github.com/apache/airflow/pull/33281/commits/2984194c7f3be440831c6b92d577271c9027600b

File: airflow/providers/elasticsearch/log/es_task_handler.py (hunk 4 above)
Comment (sunank200) [reply airflow/providers/elasticsearch/log/es_task_handler.py:103]: changed it: https://github.com/apache/airflow/pull/33281/commits/2984194c7f3be440831c6b92d577271c9027600b

**Review by pankajkoti — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Review by utkarsharma2 — APPROVED** [review]

**Comment by eladkal** [comment]
I may be missing something here but...
This PR added backward compatibility for `elasticsearch<8`
yet min version is 8.10
https://github.com/apache/airflow/blob/064fc2b7751a44e37ccce97609cff7c496098e56/generated/provider_dependencies.json#L375

so either we need to relax the min version or we should remove the backward compatibility code cc @dstandish

**Comment by potiuk** [comment]
I'd be for removing it. People can use previous provider versions for ES < 8 if they need

**Comment by eladkal** [comment]
https://github.com/apache/airflow/pull/35707 raised to remove the functionality added by this PR
